<a href="https://colab.research.google.com/github/dabarriosc/AlgorithmsUN2023I/blob/main/dabarrioscGroup1Lab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB
!./APlusB

5 8
13

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
-5 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
4

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt 

In [ ]:
!grep -xvFf output1.txt output2.txt 

4


In [ ]:
!grep -xvFf output2.txt output1.txt 

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

4 5
9


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

6
12 20 35 15
11 6
700


### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

90


### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [ ]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


### Test 3  stress tests 20  times  random n in [2, 20] 

In [ ]:
%%writefile test3.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>


using namespace std;

long long MaxPairwiseProduct(const vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    srand(time(NULL)); // inicializar semilla aleatoria
    for (int i = 0; i < 20; i++) {
        int n = rand() % 19 + 2; // tamaño aleatorio del arreglo entre [2, 20]
        vector<int> numbers(n);
        for (int j = 0; j < n; j++) {
            numbers[j] = rand() % 1000; // números aleatorios entre [0, 999]
        }
        cout << MaxPairwiseProduct(numbers) << endl;
    }
    return 0;
}



Writing test3.cpp


In [ ]:
!g++ test3.cpp -o test3
!./test3

560490
738826
840745
699030
483360
943680
611184
849243
840875
961324
930204
377274
642048
984015
919884
446367
906876
406608
208725
428198


### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [ ]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Writing Test2Write.cpp


In [ ]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

39999800000

real	2m20.463s
user	2m19.298s
sys	0m0.071s


##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line 

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;                         //1
    std::vector<int> numbers(n);        //n                                    
    for (int i = 0; i < n; ++i) {       //1+2n                                    
        std::cout << numbers[i];        //n                                     
    }

    std::cout << "\n";                  //1     
    return 0;                                                              
}                                       //1+n+1+2n+n+1

Writing test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Writing max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182
2
2
2
2
2
2
2
2
1
0
80003672077


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

### Efficient solution

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    execints++;
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
    }

    std::cout << "Cambios hechos: " << m << "\n";
                                                                            // -------------- 
    return max_product;                                                     // 5 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    auto t1 = std::chrono::high_resolution_clock::now();
    long long execints =0LL;                                 

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);                           //1

    std::cout << "Execints: " << execints << "\n";                          

    auto t2 = std::chrono::high_resolution_clock::now(); 

    // cálculo de la duración de la ejecución
    auto duracion = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1);        

    std::cout << "Tiempo empleado: " << duracion.count() << " microsegundos en ejecutarse." << std::endl; 

    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product_long_execints.cpp


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt

Cambios hechos: 30
Execints: 289
Tiempo empleado: 81 microsegundos en ejecutarse.
Cambios hechos: 5
Execints: 116
Tiempo empleado: 77 microsegundos en ejecutarse.


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input3.txt

Execints: 80003672077
Tiempo empleado: 207192200 microsegundos en ejecutarse.


20 stress tests

In [ ]:
%%writefile test3.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <chrono>


using namespace std;

long long MaxPairwiseProduct(const vector<int>& numbers,long long execints) {
    long long max_product = 0LL;
    execints++;

	  long long aux_product = 0LL;
    execints++;

    int n = numbers.size();
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {
        execints += 2;

        execints++;
        for (int second = first + 1; second < n; ++second) {
            execints += 2;

			      aux_product =  (long long) numbers[first] * numbers[second];
            execints++;

            execints++;
            if ( max_product < aux_product)  {
				        max_product = aux_product;
                execints++;
            }
        }
    }

    execints++;
    return max_product;
}

int main() {
    auto t1 = std::chrono::high_resolution_clock::now();
    long long execints =0LL;

    srand(time(NULL)); // inicializar semilla aleatoria
    execints++;

    execints++;
    for (int i = 0; i < 20; i++) {
        execints += 2;

        int n = rand() % 19 + 2; // tamaño aleatorio del arreglo entre [2, 20]
        execints++;

        vector<int> numbers(n);
        execints += n;

        execints++;
        for (int j = 0; j < n; j++) {
            execints += 2;

            numbers[j] = rand() % 1000; // números aleatorios entre [0, 999]
            execints++;
        }
        //cout << MaxPairwiseProduct(numbers) << endl;
        execints++;
    }

    cout << "Instrucciones: " << execints << endl;

    auto t2 = std::chrono::high_resolution_clock::now(); 

    // cálculo de la duración de la ejecución
    auto duracion = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1);        

    std::cout << "Tiempo empleado: " << duracion.count() << " microsegundos en ejecutarse." << std::endl; 
    return 0;
}


Overwriting test3.cpp


In [ ]:
!g++ test3.cpp -o test3
!./test3

Instrucciones: 954
Tiempo empleado: 51 microsegundos en ejecutarse.


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
print("Instrucciones input1: " + str(9+(7*10)+(2*9*10)+9+5+4+3+2+2+2+2+1))

Instrucciones input1: 289


In [ ]:
print("Instrucciones input2: " + str(9+(7*6)+(2*5*6)+2+1+2))

Instrucciones input2: 116


In [ ]:
print(str(9+(7*200000)+(2*199999*200000)) + " <= Instrucciones input3  <= " + str(9+(7*200000)+(2.5*199999*200000)))

80001000009 <= Instrucciones input3  <= 100000900009.0


Calculate estimated time by instruction and number of instructions by second

In [ ]:
print(str(0.000081/289) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son: "+ str(1/(0.000081/289)))

2.8027681660899653e-07 Segundos por instruccion
La cantidad de instrucciones por segundo son: 3567901.2345679016


In [ ]:
print(str(0.000077/116) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son: "+ str(1/(0.000077/116)))

6.637931034482759e-07 Segundos por instruccion
La cantidad de instrucciones por segundo son: 1506493.5064935065


In [ ]:
print(str(207.192200/80003672077) + "Segundos por instruccion")
print("La cantidad de instrucciones por segundo son: "+ str(1/(207.192200/80003672077)))

2.58978362643888e-09Segundos por instruccion
La cantidad de instrucciones por segundo son: 386132644.3611294


### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line 

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>

int main() {                                                 //N inst exec by line executed 
    auto t1 = std::chrono::high_resolution_clock::now();
    long long execints =0LL;                                

    long long n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    int max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+(n-1)n
            execints += 2;     

            max_product = std::max(max_product,numbers[first]*numbers[second]); //(n-1)n/2
            execints++;

        }
    }

    std::cout << "Execints: " << execints << "\n";                            

    auto t2 = std::chrono::high_resolution_clock::now(); 

    // cálculo de la duración de la ejecución
    auto duracion = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1);        

    std::cout << "Tiempo empleado: " << duracion.count() << " microsegundos en ejecutarse." << std::endl; 

    return 0;                                                // -------------- 
}                                                           //6 + 7n + 1.5n(n-1)

Overwriting max_pairwise_product_long_execints.cpp


Test with inputs 1 and 2


 

In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt

Execints: 211
Tiempo empleado: 429 microsegundos en ejecutarse.
Execints: 93
Tiempo empleado: 76 microsegundos en ejecutarse.


Test and time it with input 3

In [ ]:
!./max_pairwise_product_long_execints < input3.txt

Execints: 60001100006
Tiempo empleado: 201507926 microsegundos en ejecutarse.


20 stress tests 

In [ ]:
%%writefile test3.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>

int main() {                                                 //N inst exec by line executed 
    auto t1 = std::chrono::high_resolution_clock::now();
    long long execints =0LL;                                
                                                    
    
    srand(time(NULL));                                                   //1
    execints++;          

    execints++;                                                           //1
    for(int i=0; i<20; i++){                                              //41
        execints += 2;

        int n = rand() % 19 + 2;                                          //1
        execints++;

        vector<int> numbers(n);                                           //1
        execints += n;

        execints++;                                                        
        for (int j = 0; j < n; j++) {                                      //1+2n
            execints += 2;

            numbers[j] = rand() % 1000;                                    //n
            execints++;
        }

        int max_product = 0LL;                                             //1
        execints++;

        long long aux_product = 0LL;                                       //1
        execints++;

        execints++;                                                                         
        for (int first = 0; first < n; ++first) {                             //1+2n  
          execints += 2;

          execints++;                                    
          for (int second = first + 1; second < n; ++second) {                 //n+(n-1)n
              execints += 2;     

              max_product = std::max(max_product,numbers[first]*numbers[second]); //(n-1)n/2
              execints++;

          }
        }
    }                                                  
    
    std::cout << "Execints: " << execints << "\n";                            

    auto t2 = std::chrono::high_resolution_clock::now(); 

    // cálculo de la duración de la ejecución
    auto duracion = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1);        

    std::cout << "Tiempo empleado: " << duracion.count() << " microsegundos en ejecutarse." << std::endl; 

    return 0;                                                // -------------- 
}          

In [ ]:
!g++ test3.cpp -o test3
!./test3

Instrucciones: 914
Tiempo empleado: 66 microsegundos en ejecutarse.


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
print("Instrucciones input1: " + str(6+(7*10)+(1.5*10*9)))

Instrucciones input1: 211.0


In [ ]:
print("Instrucciones input2: " + str(6+(7*6)+(1.5*6*5)))

Instrucciones input2: 93.0


In [ ]:
print("Instrucciones input2: " + str(6+(7*200000)+(1.5*200000*199999)))

Instrucciones input2: 60001100006.0


Calculate estimated time by instruction and number of instructions by second

In [ ]:
print(str(0.000429/211) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son en Input1: "+ str(1/(0.000429/211)))

2.033175355450237e-06 Segundos por instruccion
La cantidad de instrucciones por segundo son en Input1: 491841.49184149184


In [ ]:
print(str(0.000076/93) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son en Input2: "+ str(1/(0.000076/93)))

8.172043010752689e-07 Segundos por instruccion
La cantidad de instrucciones por segundo son en Input2: 1223684.2105263157


In [ ]:
print(str(201.507926/60001100006) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son en Input3: "+ str(1/(201.507926/60001100006)))

3.358403862260018e-09 Segundos por instruccion
La cantidad de instrucciones por segundo son en Input3: 297760496.06108296


# Codeforces problem https://codeforces.com/contest/1619/problem/A 

Solve the problem in Colab, submit the solution to Codeforces and add a screenshoot of the submission.

In [ ]:
n = int(input())

for i in range(n):
    s = input()
    if len(s) % 2 == 0:
        half = len(s) // 2
        if s[:half] == s[half:]:
            print("YES")
        else: 
          print("NO")
    else:
      print("NO")

10
a
NO
aa
YES
aaa
NO
aaaa
YES
abab
YES
abcabc
YES
xxyy
NO
xyyx
NO
abacaba
NO
xyxy
YES
